File: src/improvement/benchmark.py
"""Performance benchmark framework — track operations over time.

Defines benchmarks, runs them, stores results, detects regressions.
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import asdict, dataclass
from pathlib import Path


@dataclass
class BenchmarkResult:
    name: str
    duration_ms: float
    timestamp: float
    commit: str = ""
    metadata: dict | None = None


@dataclass
class RegressionAlert:
    benchmark: str
    baseline_ms: float
    current_ms: float
    regression_pct: float
    threshold_pct: float


def run_benchmark(name: str, fn: callable, iterations: int = 5) -> BenchmarkResult:
    """Run a benchmark function multiple times and return median duration."""
    durations = []
    for _ in range(iterations):
        start = time.perf_counter()
        fn()
        elapsed = (time.perf_counter() - start) * 1000  # ms
        durations.append(elapsed)

    durations.sort()
    median = durations[len(durations) // 2]

    return BenchmarkResult(
        name=name,
        duration_ms=round(median, 3),
        timestamp=time.time(),
    )


def save_result(result: BenchmarkResult, results_dir: str) -> str:
    """Append benchmark result to JSONL file."""
    os.makedirs(results_dir, exist_ok=True)
    filepath = os.path.join(results_dir, f"{result.name}.jsonl")
    with open(filepath, "a") as f:
        f.write(json.dumps(asdict(result)) + "\n")
    return filepath


def load_history(name: str, results_dir: str, max_entries: int = 100) -> list[BenchmarkResult]:
    """Load benchmark history from JSONL file."""
    filepath = os.path.join(results_dir, f"{name}.jsonl")
    if not os.path.exists(filepath):
        return []

    results = []
    with open(filepath) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                data = json.loads(line)
                results.append(BenchmarkResult(**{
                    k: v for k, v in data.items() if k in BenchmarkResult.__dataclass_fields__
                }))
            except (json.JSONDecodeError, TypeError):
                continue

    return results[-max_entries:]


def detect_regression(
    current: BenchmarkResult,
    history: list[BenchmarkResult],
    threshold_pct: float = 10.0,
    min_history: int = 3,
) -> RegressionAlert | None:
    """Detect if current result is a regression vs historical baseline.

    Uses median of historical results as baseline.
    """
    if len(history) < min_history:
        return None

    durations = sorted(r.duration_ms for r in history)
    baseline = durations[len(durations) // 2]

    if baseline == 0:
        return None

    regression_pct = ((current.duration_ms - baseline) / baseline) * 100

    if regression_pct > threshold_pct:
        return RegressionAlert(
            benchmark=current.name,
            baseline_ms=baseline,
            current_ms=current.duration_ms,
            regression_pct=round(regression_pct, 1),
            threshold_pct=threshold_pct,
        )

    return None

File: rust/src/improvement/benchmark.rs
//! Performance benchmark framework — run benchmarks, detect regressions.

use serde::{Deserialize, Serialize};

/// Result of a single benchmark run.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BenchmarkResult {
    pub name: String,
    pub duration_ms: f64,
    pub timestamp: f64,
    #[serde(default)]
    pub commit: String,
    #[serde(default)]
    pub metadata: Option<serde_json::Value>,
}

/// Alert when a benchmark regresses.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegressionAlert {
    pub benchmark: String,
    pub baseline_ms: f64,
    pub current_ms: f64,
    pub regression_pct: f64,
    pub threshold_pct: f64,
}

/// Run a benchmark function multiple times and return median duration.
pub fn run_benchmark<F: Fn()>(name: &str, f: F, iterations: usize) -> BenchmarkResult {
    let mut durations = Vec::with_capacity(iterations);
    for _ in 0..iterations {
        let start = std::time::Instant::now();
        f();
        let elapsed = start.elapsed().as_secs_f64() * 1000.0;
        durations.push(elapsed);
    }
    durations.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let median = durations[durations.len() / 2];

    BenchmarkResult {
        name: name.to_string(),
        duration_ms: (median * 1000.0).round() / 1000.0,
        timestamp: std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs_f64(),
        commit: String::new(),
        metadata: None,
    }
}

/// Detect if current result is a regression vs historical baseline.
///
/// Uses median of historical results as baseline.
pub fn detect_regression(
    current: &BenchmarkResult,
    history: &[BenchmarkResult],
    threshold_pct: f64,
    min_history: usize,
) -> Option<RegressionAlert> {
    if history.len() < min_history {
        return None;
    }

    let mut durations: Vec<f64> = history.iter().map(|r| r.duration_ms).collect();
    durations.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let baseline = durations[durations.len() / 2];

    if baseline == 0.0 {
        return None;
    }

    let regression_pct = ((current.duration_ms - baseline) / baseline) * 100.0;

    if regression_pct > threshold_pct {
        Some(RegressionAlert {
            benchmark: current.name.clone(),
            baseline_ms: baseline,
            current_ms: current.duration_ms,
            regression_pct: (regression_pct * 10.0).round() / 10.0,
            threshold_pct,
        })
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_run_benchmark() {
        let result = run_benchmark("test_bench", || {
            let _ = (0..100).sum::<i32>();
        }, 3);
        assert_eq!(result.name, "test_bench");
        assert!(result.duration_ms >= 0.0);
    }

    #[test]
    fn test_detect_regression_insufficient_history() {
        let current = BenchmarkResult {
            name: "bench".into(),
            duration_ms: 100.0,
            timestamp: 0.0,
            commit: String::new(),
            metadata: None,
        };
        assert!(detect_regression(&current, &[], 10.0, 3).is_none());
    }

    #[test]
    fn test_detect_regression_found() {
        let current = BenchmarkResult {
            name: "bench".into(),
            duration_ms: 150.0,
            timestamp: 0.0,
            commit: String::new(),
            metadata: None,
        };
        let history: Vec<BenchmarkResult> = (0..5)
            .map(|_| BenchmarkResult {
                name: "bench".into(),
                duration_ms: 100.0,
                timestamp: 0.0,
                commit: String::new(),
                metadata: None,
            })
            .collect();
        let alert = detect_regression(&current, &history, 10.0, 3);
        assert!(alert.is_some());
        assert_eq!(alert.unwrap().regression_pct, 50.0);
    }

    #[test]
    fn test_detect_regression_within_threshold() {
        let current = BenchmarkResult {
            name: "bench".into(),
            duration_ms: 105.0,
            timestamp: 0.0,
            commit: String::new(),
            metadata: None,
        };
        let history: Vec<BenchmarkResult> = (0..5)
            .map(|_| BenchmarkResult {
                name: "bench".into(),
                duration_ms: 100.0,
                timestamp: 0.0,
                commit: String::new(),
                metadata: None,
            })
            .collect();
        assert!(detect_regression(&current, &history, 10.0, 3).is_none());
    }
}
